var/spack/repos/builtin/packages/abinit/package.py

```diff
@@ -72,29 +61,37 @@ class Abinit(AutotoolsPackage):
 
     depends_on('scalapack', when='+scalapack+mpi')
 
-    # depends_on('elpa~openmp', when='+elpa+mpi~openmp')
-    # depends_on('elpa+openmp', when='+elpa+mpi+openmp')
-
-    depends_on('fftw precision=float,double')
-    depends_on('fftw~openmp', when='~openmp')
     depends_on('fftw+openmp', when='+openmp')
```

Comment:
Did you mean fftw explicitly here?

Reply:
Yes, although it is less than ideal. For the `+openmp` variant, use threaded fftw3. Otherwise, for the serial `~openmp` variant the fftw-api virtual is used, which allows using MKL for FFT. There is still an issue with which thread model mkl was installed with. Now that I am thinking about it, on a new day, I might be able to specify the mkl libraries explicitly, as is done for fftw3, and be able to then use fftw-api in both cases. That is worth investigating.

Reply:
The idea I mentioned would let me be able to use MKL in the threaded case but there is still the problem of fftw needing to have the `openmp` variant specified. The problem is that there is no way to specify a variant of the `fftw-api` virtual package. If fftw is the provider for `fftw-api`, then for the `+openmp` variant of abinit, the `+openmp` variant of fftw is needed to get `libfftw_omp.so`. It would be better if the fftw package built all of the library types so that they would all be available for `fftw-api`. I suppose another option is to always use non-threaded FFT.

Reply:
I pushed a commit taking a different approach. I have set a `conflicts` directive for `fftw~openmp` when `abinit+openmp`. That way, the virtual `fftw-api` can be used and the mismatched case can be handled. That is still less than ideal but I think a bit better.

Reply:
I shared this neat trick in the slack appreciation channel the other day. If you stop to care about the old concretizer (i mean, we are on `develop` after all...), you may be better off with this idea:

```
    depends_on('fftw-api')
    depends_on('fftw +openmp', when='+openmp ^fftw')
```

i.e. if your package uses openmp and the fftw-api provider happens to be fftw, then depend on fftw +openmp :)

Reply:
The drawback of the conflict approach is that it may lead the concretizer to using a different provider instead of changing the default from the preferred provider

Reply:
Let me know if you want to stick to conflicts for compatibility.

Reply:
I had actually tried using that logic with the "original" concretizer but could not get it to work. That is why I went with the conflicts. It is good to know that will work with the new concretizer. I have not had a chance to play with that much.
>The drawback of the conflict approach is that it may lead the concretizer to using a different provider instead of changing the default from the preferred provider

I do not see this with the original concretizer so I assume that you are referring to the behavior with the new concretizer.

I think many, perhaps most, people use the develop branch so until the new concretizer is the default I think it is best to go with the `conflicts` directives. I will add a `TODO` note to change this once the concretizer default changes.

Reply:
First of all, thanks to @glennpj for porting the abinit package to v9.4.2.
Sorry about the big changes in the build system API. I imagine that supporting both v8 and v9 with the same recipe required a lot of extra work.

I haven't tried the new recipe but since I see that MKL is mentioned in the PR, I would like to point out that one has to be very careful when compiling Abinit with openMP + MKL and FFTW3.

MKL indeed provides BLAS/LAPACK/Scalapack as well as wrappers implementing the FFTW3 API 
that is actually calling the native intel DFTI library.
Note however that the MKL-FFTW3 wrappers are not fully thread safe.
See, for instance. <https://scc.ustc.edu.cn/zlsc/sugon/intel/mkl/mkl_manual/GUID-D17B3AB4-BD4E-4652-94A7-BAD4130CCB4A.htm>

What is even worse is that these MKL wrappers are bundled in the same libs providing BLAS/LAPACK.
As a consequence, one should avoid linking ABINIT with MKL for the BLAS/LAPACK part and the **true** FFTW3 library  for the FFT part else we end up with multiple defined symbols passed to the linker (undefined behavior).

In other words, FFTW3 and MKL are mutually exclusive and this is especially true when OpenMP support is activated as in our FFT machinery we assume a tread-safe FFT library.

Reply:
> As a consequence, one should avoid linking ABINIT with MKL for the BLAS/LAPACK part and the **true** FFTW3 library for the FFT part else we end up with multiple defined symbols passed to the linker (undefined behavior).
> 
> In other words, FFTW3 and MKL are mutually exclusive and this is especially true when OpenMP support is activated as in our FFT machinery we assume a tread-safe FFT library.

That combination could be addressed with another conflicts statement.

Reply:
> Note however that the MKL-FFTW3 wrappers are not fully thread safe.

I guess even using `mkl+openmp` does not help for the FFT part. Should we disable `+openmp` when using MKL?